### PR TITLE
Continue to fire timers for heartbeat during thread pool shutdown

### PR DESCRIPTION
**Pull request by owenmjones**
## Description

This extends the fix to the issue of broker heartbeats not being sent on worker shutdown, reported in Issue #5998 and fixed by @weetster in PR #9986 the context of the default `prefork` pool type.  Here this fix is applied to the `threads` pool type, which still exhibits the bug, and an integration test is added to validate the fix.

I have refactored the code as far as possible to avoid code duplication, and have used pytest fixtures to apply the same test to both `prefork` and `threads` pool types (this test refactor is the only part of the PR for which I used AI assistance, with Claude Opus helping me with the fixture syntax and debugging a thread-locking problem in the test).

It is possible that the issue persists in the context of `eventlet` and/or `gevent` async pool types, but my understanding of these pool types is insufficient to allow me to propose a fix for them.

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
did you check this PR? https://github.com/celery/celery/pull/10151 

the PR you are extending, had some regression. so we have to consider this carefully. @ChickenBenny it would be great if you can also have a look into this

File: celery/concurrency/prefork.py [hunk 1]
```diff
@@ -3,19 +3,16 @@
...
 from billiard.common import REMAP_SIGTERM, TERM_SIGNAME
 from billiard.pool import CLOSE, RUN
 from billiard.pool import Pool as BlockingPool
-from kombu.asynchronous import get_event_loop
 
 from celery import platforms, signals
 from celery._state import _set_task_join_will_block, set_default_app
 from celery.app import trace
-from celery.concurrency.base import BasePool
+from celery.concurrency.base import AsyncPoolShutdownMixin, BasePool
```
[the remark this answers is not in the document]
Comment (owenmjones) [reply celery/concurrency/prefork.py:15]: Fixed

File: celery/concurrency/base.py [hunk 2]
```diff
@@ -178,3 +180,63 @@ def active(self):
...
+                target=fire_timers_loop,
+                daemon=True,
+                name=f"{pool_type}-timer-shutdown",
+            )
+            timer_thread.start()
+
+            return shutdown_event, timer_thread
+
+    @classmethod
+    def shutdown_with_timer_loop(cls, *_, pool_type: str, shutdown_function: Callable):
```
[the remark this answers is not in the document]
Comment (owenmjones) [reply celery/concurrency/base.py]: Fixed

File: celery/concurrency/thread.py [hunk 3]
```diff
@@ -17,6 +19,8 @@
     # trace_task_ret.
     TargetFunction = Callable[..., Any]
 
+logger = get_logger(__name__)
```
[the remark this answers is not in the document]
Comment (owenmjones) [reply celery/concurrency/thread.py]: Forgot to clean this up when adding the second commit. I've also removed the unused `logger` definition from `celery/concurrency/prefork.py`.

**Comment by owenmjones** [comment]
> did you check this PR? #10151
> 
> the PR you are extending, had some regression. so we have to consider this carefully

@auvipy no, I didn't see that, thanks for bringing it to my attention. If my understanding is correct, the fix proposed here works with that change since the `asynloop` behaviour from #9986 is preserved in the normal shutdown case (and this code is not executed if any other `Exception` is raised). The changes made in this PR are all "downstream" of the `asynloop` exit, and pertain to the case where the event loop raises `WorkerShutdown`.

It's probably best if someone else verifies the shutdown behaviour of both `prefork` and `threads` pools on this branch when using Redis Sentinel, since I'm not really familiar with that broker.

**Comment by auvipy** [comment]
it's alright

**Comment by ChickenBenny** [comment]
Great work on the refactoring! However, I have some concerns regarding the thread safety of this approach when applied to the threads pool.

In the prefork pool, we've already seen issues (like the regression in #10151) where the background Timer thread tries to send heartbeats while the main thread is simultaneously tearing down the broker connection. This leads to race conditions, such as `RuntimeError` or broken pipes.

If we apply this to the threads pool, the risk seems to multiply. During a graceful shutdown, we would have a three-way contention:
1. The Main Thread initiating the shutdown sequence and cleaning up connection pools.
2. Active Worker Threads finishing their tasks and concurrently writing results or sending ACKs to the broker.
3. This new Timer Thread concurrently firing heartbeats over the same shared broker connection.

Since everything shares the same memory space in the threads pool, we are hitting the connection management logic from multiple angles while it's actively being destroyed.

Have we verified if the underlying broker transports (especially `kombu` with `redis-py` or `Sentinel`) can safely handle this level of concurrent access during the shutdown phase?1

**Comment by owenmjones** [comment]
@ChickenBenny I've just tested this locally using the `redis` broker, running in `--pool=threads` mode with `--concurrency=4` and launching multiple long-running tasks. I added print statements at various points so we can see what's happening: here is the output.
```
...

[2026-03-11 09:52:56,432: WARNING/MainProcess] Sending heartbeat event worker-heartbeat

[2026-03-11 09:52:58,434: WARNING/MainProcess] Sending heartbeat event worker-heartbeat

[2026-03-11 09:53:00,440: WARNING/MainProcess] Sending heartbeat event worker-heartbeat

[2026-03-11 09:53:00,639: INFO/MainProcess] Task myproject.tasks.long_running_task[753288b2-3393-44e2-99db-202a04263734] received

[2026-03-11 09:53:00,640: WARNING/MainProcess] Started long-running task 1

[2026-03-11 09:53:00,642: INFO/MainProcess] Task myproject.tasks.long_running_task[e42e144d-fa37-4690-bf42-60043c08c9c5] received

[2026-03-11 09:53:00,643: WARNING/MainProcess] Started long-running task 2

[2026-03-11 09:53:00,645: INFO/MainProcess] Task myproject.tasks.long_running_task[78645c96-59c7-4981-96c9-6693d97608e2] received

[2026-03-11 09:53:00,646: WARNING/MainProcess] Started long-running task 3

[2026-03-11 09:53:00,647: INFO/MainProcess] Task myproject.tasks.long_running_task[d8b6dd3e-74ca-4a3a-a4f7-193545a080f6] received

[2026-03-11 09:53:00,648: WARNING/MainProcess] Started long-running task 4

[2026-03-11 09:53:02,446: WARNING/MainProcess] Sending heartbeat event worker-heartbeat

[2026-03-11 09:53:04,453: WARNING/MainProcess] Sending heartbeat event worker-heartbeat

[2026-03-11 09:53:06,456: WARNING/MainProcess] Sending heartbeat event worker-heartbeat


worker: Warm shutdown (MainProcess)

[2026-03-11 09:53:07,334: WARNING/MainProcess] Stopping Heart

[2026-03-11 09:53:07,334: WARNING/MainProcess] Sending heartbeat event worker-offline

[2026-03-11 09:53:07,340: WARNING/MainProcess] Shutting down thread TaskPool

[2026-03-11 09:53:07,341: WARNING/MainProcess] Starting timer thread

[2026-03-11 09:53:07,341: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:07,843: WARNING/MainProcess] Firing kombu hub timers from timer thread

... (previous message repeated every 0.5s)

[2026-03-11 09:53:21,282: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:21,552: WARNING/MainProcess] Finished long-running task 1

[2026-03-11 09:53:21,594: INFO/MainProcess] Task myproject.tasks.long_running_task[753288b2-3393-44e2-99db-202a04263734] succeeded in 21.04304429700005s: None

[2026-03-11 09:53:21,783: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:22,283: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:22,554: WARNING/MainProcess] Finished long-running task 2

[2026-03-11 09:53:22,569: INFO/MainProcess] Task myproject.tasks.long_running_task[e42e144d-fa37-4690-bf42-60043c08c9c5] succeeded in 22.015423068999553s: None

[2026-03-11 09:53:22,784: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:23,284: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:23,557: WARNING/MainProcess] Finished long-running task 3

[2026-03-11 09:53:23,574: INFO/MainProcess] Task myproject.tasks.long_running_task[78645c96-59c7-4981-96c9-6693d97608e2] succeeded in 23.017729862000124s: None

[2026-03-11 09:53:23,785: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:24,285: WARNING/MainProcess] Firing kombu hub timers from timer thread

[2026-03-11 09:53:24,560: WARNING/MainProcess] Finished long-running task 4

[2026-03-11 09:53:24,577: INFO/MainProcess] Task myproject.tasks.long_running_task[d8b6dd3e-74ca-4a3a-a4f7-193545a080f6] succeeded in 24.018544253999607s: None

[2026-03-11 09:53:24,786: WARNING/MainProcess] Stopped timer thread

[2026-03-11 09:53:24,786: WARNING/MainProcess] Shut down thread TaskPool
```
So we can see the `MainProcess` and `timer` threads start and stop as expected, and that the tasks are indeed run concurrently and execute successfully in four worker threads.

If you have more specific concerns about a particular scenario which may cause problems (e.g. triggering a broker connection loss during shutdown) I can try to reproduce it.

**Comment by ChickenBenny** [comment]
Thanks for taking the time to test this and sharing the detailed logs, @owenmjones! I really appreciate the effort. It’s good to see that the timer thread spins up and shuts down as expected in a happy-path scenario.

However, from a concurrency perspective, a simple local test with 4 tasks doesn't fully guarantee thread safety, especially when we are dealing with shared connection states. Here is why:

While Python has the GIL (Global Interpreter Lock), the GIL is intentionally released during I/O operations (like socket writes to Redis or network reads). This means that during the graceful shutdown phase, the following actions are executing **truly concurrently** at the OS and network level:
1. **Timer Thread:** Actively requesting a connection from the pool to write a heartbeat to the socket.
2. **Worker Threads:** Requesting connections to write task results and ACKs.
3. **Main Thread:** Potentially tearing down the hub, closing the connection pools, and destroying the socket file descriptors.

The local test passes because the timing hasn't perfectly collided yet. The real edge case (and the root of many `RuntimeError` or `BrokenPipeError` issues we see in production) happens when the Timer Thread tries to write a heartbeat using a connection/socket that the Main Thread is exactly in the middle of tearing down.

To truly validate this approach for the `threads` pool, we shouldn't rely solely on execution logs. We need to look deeper into the underlying implementation of `kombu.transport.redis` and the `redis-py` connection pool. We need to answer: **Are Kombu's asynchronous hub and the Redis connection pool strictly thread-safe when connections are being actively destroyed by another thread?**

If the underlying transport isn't designed to handle concurrent teardown and writes from different threads, applying this mixin to the `threads` pool might introduce rare but severe race conditions in production.

**Comment by owenmjones** [comment]
@ChickenBenny thanks for providing valuable extra context! Is there a reason you mention only `redis`/`redis-py` and not `rabbitmq`? Is `rabbitmq` thread-safe by construction whereas `redis` isn't?

**Comment by ChickenBenny** [comment]
@owenmjones The main reason I mentioned Redis is simply because I recently fixed a concurrency issue related to it in #10151, so the context is very fresh in my mind! 😅

With `redis-py`, the connection pool is indeed thread-safe for standard operations. However, edge cases often arise during lifecycle state transitions, such as a **graceful shutdown**. 

For example: when the main process attempts to tear down the pool and close all connections during a graceful shutdown, you might simultaneously have a worker thread trying to fire a task and another thread trying to store a result. Both threads are competing to acquire a connection from the shared pool. If no connections are available, they might attempt to instantiate new ones while the main process is in the middle of destroying the pool. Because these threads share the exact same connection pool instance without a strict lifecycle lock during teardown, it leads to a race condition. The connections can't be cleaned up properly, ultimately throwing a `RuntimeError`.

So, it's less about the protocol itself and more about how the shared pool's state is managed during these tricky lifecycle events!

**Comment by auvipy** [comment]
is this in merge able state now?

**Comment by owenmjones** [comment]
@auvipy no, we either need someone with more expertise in the broker transports to look into this or else we need to come up with an alternative approach to guarantee thread-safety. See the comments by @ChickenBenny: as it stands there is too much risk that this would introduce a serious regression due to concurrency issues in some edge cases.
